### DISCUSSION TITLE
Can incremental snapshots be restored at all?

### DISCUSSION THREAD

**Discussion by 2fst4u** [vmware-tanzu/velero#8554] [Community support Q&A] [answered]
I'm a bit confused by the restore process. I was hoping to be able to schedule backups and have snapshots be taken, allowing the CSI snapshots to be restored in the regular way of creating a PVC by referencing the snapshot.

I understand velero now removes snapshots once a backup is done so that it's not accidentally removed if the namespace is deleted, which I understand.

But is there now no way for a snapshot to be restored? Is the only way to restore data to do a full PVC restore?

My concern there is I really have no way of knowing my backups are working because the only way to restore them to test them is to destroy the PVC, to restore over it - unless you have another cluster to restore into, which I don't have.

Am I missing something?

**Comment by kaovilai** [comment]
> is there now no way for a snapshot to be restored? 

No. [VolumeSnapshotContent remains in the cluster. It is intentionally disconnected from VolumeSnapshot so it lives past namespace deletion.
](https://velero.io/docs/v1.15/csi/#:~:text=VolumeSnapshotContent%20objects%20created%20during%20a%20velero%20backup%20that%20are%20dangling%2C%20unbound%20to%20a%20VolumeSnapshot%20object%2C%20will%20be%20discovered%2C%20using%20labels%2C%20and%20deleted%20on%20backup%20deletion.)
[VSC will only get deleted when backup gets deleted](https://velero.io/docs/v1.15/csi/#:~:text=VolumeSnapshots%20created%20by,the%20storage%20provider.)

On restore velero create a volumesnapshot object referencing the original volumesnapshot**content**  and will be able to create a PVC using VolumeSnapshot as source.

>  I really have no way of knowing my backups are working

Check there are volumesnapshotcontents with your backup name in its labels/annotations.

**Reply by 2fst4u** [reply]
There are volumesnapshotcontents indeed, forgive my confusion but I thought those were incremental and therefore wouldn't quite work to be restored into a PVC? I may have read too many documents and got myself confused.

**Reply by kaovilai** [reply]
Every volumesnapshot has its associated volumesnapshotcontents. Incrementalness depends on csi driver/cloud provider and not whether its VS or VSC. A VS always has associated VSC.

**Reply by 2fst4u** [reply]
But the volumesnapshots get deleted right?

**Reply by kaovilai** [reply]
VolumeSnapshots can be deleted without VSC getting deleted.
VS can be recreated from VSC and used later to restore.

**Reply by kaovilai** [reply]
https://github.com/vmware-tanzu/velero/blob/9dcfe164d8094df41353866378fb91b741bfe094/pkg/exposer/csi_snapshot.go#L143-L153
Here we first make sure VSC is retained, then VS is deleted.
Thus you can simulate disaster by deleting NS of workload without VSC deletion and restore the NS later.

**Reply by kaovilai** [reply]
If we [had **NOT** retained VSC manually](https://github.com/vmware-tanzu/velero/discussions/8554#discussioncomment-11655991:~:text=retained%2C%20err%20%3A%3D%20csi.RetainVSC(ctx%2C%20e.csiSnapshotClient%2C%20vsc)) in this case, then yes, a VS deletion = VSC deletion = poof no restore

**Reply by 2fst4u** [reply]
I get that the VSC is retained which is good, but then what I'm confused about is don't you need the VS to restore a PVC from? Which has now since been deleted?

**Reply by kaovilai** [reply]
There is a way to delete VS without deleting VSC.

VS itself has no data but references VSC which has the data.

**Reply by 2fst4u** [reply]
I get that the VS can be deleted without the VSC but I'm just trying to wrap my head around how you get the data. Manually making a VS to reference it and then manually making a PVC to reference that doesn't seem very user friendly, I'm just trying to figure out if there is an easy way.

**Comment by kaovilai** [comment, marked as answer]
> the only way to restore them to test them is to destroy the PVC, to restore over it

Not true.

You can create VolumeSnapshot pointing to existing VolumeSnapshotContent.
You can then create PVC with volumeSource set to the VolumeSnapshot.

**Reply by 2fst4u** [reply]
Ah ok, so it's a bit of manual intervention required. Is there an easier way to do it than creating a VS manifest and applying it to the cluster? Or is this really the only way to do it?

**Reply by kaovilai** [reply]
idk.. an alias/bash func that does the same thing? The limitation is with kube. You would need to change kube to allow PVC dataSource from VSC then you can skip VS creation.

> [The `dataSourceRef` field may contain different types of objects, while the `dataSource` field only allows PVCs and VolumeSnapshots.](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#volume-snapshot-and-restore-volume-from-snapshot-support:~:text=The%20dataSourceRef%20field%20may%20contain%20different%20types%20of%20objects%2C%20while%20the%20dataSource%20field%20only%20allows%20PVCs%20and%20VolumeSnapshots.)

**Reply by 2fst4u** [reply]
All good I just wanted to make sure I wasn't missing something. Thank you for the info, much appreciated.
